azurerm/resource_arm_front_door.go

```diff
@@ -0,0 +1,1460 @@
+package azurerm
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/Azure/azure-sdk-for-go/services/preview/frontdoor/mgmt/2019-04-01/frontdoor"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/azure"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/response"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/tf"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/frontdoor/helper"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/frontdoor/validate"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func resourceArmFrontDoor() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmFrontDoorCreateUpdate,
+		Read:   resourceArmFrontDoorRead,
+		Update: resourceArmFrontDoorCreateUpdate,
+		Delete: resourceArmFrontDoorDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validate.FrontDoorName,
+			},
+
+			"cname": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"friendly_name": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"load_balancer_enabled": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+			},
+
+			"enforce_backend_pools_certificate_name_check": {
+				Type:     schema.TypeBool,
+				Required: true,
+			},
+
+			"location": azure.SchemaLocation(),
+
+			"resource_group_name": azure.SchemaResourceGroupNameDiffSuppress(),
+
+			"routing_rule": {
+				Type:     schema.TypeList,
+				MaxItems: 100,
+				Required: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"name": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validate.BackendPoolRoutingRuleName,
+						},
+						"enabled": {
+							Type:     schema.TypeBool,
+							Optional: true,
+							Default:  true,
+						},
+						"accepted_protocols": {
+							Type:     schema.TypeList,
+							Required: true,
+							MaxItems: 2,
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+								ValidateFunc: validation.StringInSlice([]string{
+									string(frontdoor.HTTP),
+									string(frontdoor.HTTPS),
+								}, false),
+							},
+						},
+						"patterns_to_match": {
+							Type:     schema.TypeList,
+							Required: true,
+							MaxItems: 25,
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+							},
+						},
+						"frontend_endpoints": {
```

Comment:
should we name this differently then the top level `frontend_endpoints` to make the difference cleaR?

Reply:
Not sure, since they are the same thing. The frontend endpoints is how you associate the frontend endpoint to this routing rule. Hopefully my update to the docs makes this more clear.

Reply:
...this is a very odd resource since it is self referencing...